SourceX/controls/plrctrls.cpp

```diff
@@ -203,6 +213,75 @@ void CheckMonstersNearby()
 	}
 }
 
+void FindMeleeTarget()
+{
+	bool visited[MAXDUNX][MAXDUNY] = { 0 };
+	int maxSteps = 25;
+	int rotations = 5;
+	int x = plr[myplr]._px;
+	int y = plr[myplr]._py;
+
+	std::list<bfsNode> queue;
+
+	visited[x][y] = true;
+	queue.push_back({ x, y, 0 });
+
+	while (!queue.empty()) {
+		bfsNode node = queue.front();
+		queue.pop_front();
+
+		for (int i = 0; i < 8; i++) {

```

Comment:
Currently we queue the nodes in the distance order but we're trying to find the minimum by "roteryDistance". If we queued the nodes in the "roteryDistance" order instead, we could abort after the first monster instead of having to do a full BFS every time. Not sure how hard that would be.

Reply:
The BFS does abort at the level of the first monster, but maybe your right that we could abort even ealier

Reply:
Where does it abort? (might need Dijsktra, depending on roteryDistance properties).

Reply:
maxSteps is set to current depth (line 251) meaning no more nodes will be added and the next level will skip (258), but it will finish the current sweep to check for any monsters in a more optimal direction at same level.

Looking at it again I see that line 258 should be moved up before `PosOkPlayer` to avoid it finding monsters at max+1.

This being my first venture in to path finding I'm a little unsure of things, but from what I can tell this approach would qualify as a Dijsktra map, but with the added benefit (in our case) that we stop generating the map once we find the minimum distance to any given monster.

`roteryDistance` doesn't add any actual cost to movement, but it allows us to assume the players intent instead of just picking any random monster at the appropriate distance, so that is why it's only applied selectively once we find was the closest enemies are. Of-cause we could also add some more bias based on player direction, but at this point I'm not sure how much would be sensible.

Hope this all makes sense :)

Reply:
OK this all makes sense!

> maxSteps is set to current depth (line 251) meaning no more nodes will be added and the next level will skip (258),

Ah. Please add a comment!

Reply:
I think it's `rotaryDistance` :)